homework/hw1/task2/task2/Bwt.cs

```diff
@@ -0,0 +1,98 @@
+        var suffixArray = GetSuffixPositionsArray(textForRotations);
+        var resultText = new StringBuilder(encodedText.Length);
+        var endPosition = 0;
+
+        for (var i = 0; i < textForRotations.Length; i++)
+        {
+            if (suffixArray[i] >= encodedText.Length)
+            {
+                continue;
+            }
+
+            var index = suffixArray[i] - 1;
+            if (index == -1)
+            {
+                index = encodedText.Length - 1;
+                endPosition = i / 2;
+            }
+
+            resultText.Append(encodedText[index]);
+        }
+
+        return (resultText.ToString(), endPosition);
+    }
+
+    /// <summary>
+    /// Reverse Burrows–Wheeler Transformation for a string.
+    /// </summary>
+    /// <param name="decodedText">Decoded text.</param>
+    /// <param name="endPosition">End of string position after encoding.</param>
+    /// <returns>Decoded text.</returns>
+    public static string ReverseTransformation(string decodedText, int endPosition)
+    {
+        var result = new StringBuilder(decodedText.Length);
+        var symbols = string.Join("", decodedText
+            .Distinct()
+            .OrderBy(symbol => symbol));
+        var countSymbolsArray = new int[symbols.Length];
+        foreach (var symbol in decodedText)
+        {
+            countSymbolsArray[symbols.IndexOf(symbol)] += 1;
+        }
+
+        var charPositionsInSortArray = new int[decodedText.Length];
+        for (var j = 0; j < decodedText.Length; j++)
+        {
+            charPositionsInSortArray[j] = countSymbolsArray.Take(
+                                                  symbols
+                                                      .IndexOf(decodedText[j]))
+                                              .Sum() +
+                                          decodedText
+                                              .Take(j)
+                                              .Count(x => x == decodedText[j]
+                                              );
+        }
+
+        var currentPosition = endPosition;
+        while (result.Length != decodedText.Length)
+        {
+            result.Insert(0, decodedText[currentPosition]);
+            currentPosition = charPositionsInSortArray[currentPosition];
+        }
+
+        return result.ToString();
+    }
+
+    /// <summary>
+    /// Getting array of suffixes' positions after sorting them lexicographically.
+    /// </summary>
+    /// <param name="text">Text for getting suffixes from.</param>
+    /// <returns>Array of lexicographically sorted suffixes' positions.</returns>
+    private static int[] GetSuffixPositionsArray(string text)
+    {
+        var suffixPositionsArray = Enumerable
+            .Range(0, text.Length)
+            .ToArray();
+        Array.Sort(suffixPositionsArray,
+            (i, j) => string.Compare(text[i..], text[j..], StringComparison.Ordinal));
```

Comment:
```suggestion
            (i, j) => string.Compare(text[i..] + text[0..i], text[j..] + text[0..j], StringComparison.Ordinal));
```
Вместо удвоения длины строки можно было здесь строить сдвиги (так эффективнее, чем удваивать строку). 

А вообще, по-хорошему, нужно анализировать сортировать суффиксы исходной строки с добавленным символом конца входа.
Так как добавлять символ не стоит по условию задачи, нужно эмулировать наличие этого символа в строке (например, учитывая длину суффикса).